File: implementations/react/rune-host.ts
// Rune Protocol — TypeScript Host (framework-agnostic)
// Same four stores as the C# implementation.
// React is just one consumer — this runs equally in Vue, Svelte, or Node.

export type RuneType = '@' | '~' | '!' | '?';

export interface RuneBinding {
  type:        RuneType;
  identifier:  string;
  annotation?: string;
  args?:       string[];
}

export class RuneError extends Error {
  constructor(
    public readonly code: string,
    message: string,
  ) {
    super(`[${code}] ${message}`);
  }
}

// ─────────────────────────────────────────────
// State store — mutable values + computed
// ─────────────────────────────────────────────

type Subscriber = (name: string, value: unknown) => void;

export class RuneStateStore {
  private state    = new Map<string, unknown>();
  private computed = new Map<string, () => unknown>();
  private subs     = new Set<Subscriber>();

  declare(name: string, initial?: unknown)           { this.state.set(name, initial); }
  declareComputed(name: string, fn: () => unknown)   { this.computed.set(name, fn); }
  isComputed(name: string)                           { return this.computed.has(name); }

  get(name: string): unknown {
    if (this.computed.has(name)) return this.computed.get(name)!();
    if (this.state.has(name))    return this.state.get(name);
    throw new RuneError('RNE002', `Identifier '${name}' is not declared.`);
  }

  set(name: string, value: unknown): void {
    if (this.isComputed(name))
      throw new RuneError('RNE003', `Cannot sync to computed value '${name}'.`);
    if (!this.state.has(name))
      throw new RuneError('RNE002', `State '${name}' is not declared.`);

    this.state.set(name, value);
    this.subs.forEach(fn => fn(name, value));
  }

  subscribe(fn: Subscriber): () => void {
    this.subs.add(fn);
    return () => this.subs.delete(fn);
  }
}

// ─────────────────────────────────────────────
// Action registry — named async behaviors
// ─────────────────────────────────────────────

type ActionFn = (...args: unknown[]) => Promise<void> | void;

export class RuneActionRegistry {
  private actions = new Map<string, ActionFn>();

  register(name: string, fn: ActionFn) { this.actions.set(name, fn); }

  async dispatch(name: string, ...args: unknown[]): Promise<void> {
    const fn = this.actions.get(name);
    if (!fn) throw new RuneError('RNE002', `Action '${name}' is not registered.`);
    await fn(...args);
  }
}

// ─────────────────────────────────────────────
// Intent store — ? annotations, never discarded
// ─────────────────────────────────────────────

export class RuneIntentStore {
  private intent = new Map<string, string[]>();

  record(path: string, annotation: string): void {
    const list = this.intent.get(path) ?? [];
    list.push(annotation);
    this.intent.set(path, list);
  }

  for(path: string): string[]   { return this.intent.get(path) ?? []; }
  get all(): Record<string, string[]> { return Object.fromEntries(this.intent); }
}

// ─────────────────────────────────────────────
// The Host — one per screen / document
// ─────────────────────────────────────────────

export class RuneHost {
  readonly state   = new RuneStateStore();
  readonly actions = new RuneActionRegistry();
  readonly intent  = new RuneIntentStore();

  // @ read — resolves dot-notation
  read(identifier: string): unknown {
    const parts = identifier.split('.');
    if (parts.length === 1) return this.state.get(identifier);

    let value = this.state.get(parts[0]);
    for (const key of parts.slice(1)) {
      if (value == null) return null;
      value = (value as Record<string, unknown>)[key];
    }
    return value;
  }

  // ~ sync — write to mutable state
  sync(identifier: string, value: unknown): void {
    this.state.set(identifier, value);
  }

  // ! act — dispatch a named action
  act(actionName: string, ...args: unknown[]): Promise<void> {
    return this.actions.dispatch(actionName, ...args);
  }

  // ? intent — record annotation
  recordIntent(path: string, annotation: string): void {
    this.intent.record(path, annotation);
  }

  // Subscribe to any state change (used by React hooks for re-renders)
  subscribe(fn: Subscriber): () => void {
    return this.state.subscribe(fn);
  }
}

File: implementations/csharp/RuneHost.cs
// Rune Protocol — C# Host
// The host wires the four stores together and provides the surface
// that document formats or UI frameworks consume.

namespace Rune;

// ─────────────────────────────────────────────
// The Host — one per document/screen
// ─────────────────────────────────────────────

public class RuneHost
{
    public RuneStateStore    State   { get; } = new();
    public RuneActionRegistry Actions { get; } = new();
    public RuneIntentStore   Intent  { get; } = new();

    // Resolve any identifier for @ binding (read)
    public object? Read(string identifier)
        => ResolveNestedIdentifier(identifier);

    // Write for ~ binding (sync from input to state)
    public void Sync(string identifier, object? value)
        => State.Set(identifier, value);

    // Dispatch for ! binding (act)
    public Task Act(string actionName, params object?[] args)
        => Actions.Dispatch(actionName, args);

    // Record ? binding (intent — no runtime effect, preserved for tooling)
    public void RecordIntent(string elementPath, string annotation)
        => Intent.Record(elementPath, annotation);

    // Resolve dot-notation: "item.title" against a context object
    private object? ResolveNestedIdentifier(string identifier)
    {
        var parts = identifier.Split('.');

        // Simple state lookup
        if (parts.Length == 1)
            return State.Get(identifier);

        // Nested: first part is state name, rest is property path
        var root = State.Get(parts[0]);
        return parts[1..].Aggregate(root, GetProperty);
    }

    private static object? GetProperty(object? obj, string property)
    {
        if (obj is null) return null;
        var prop = obj.GetType().GetProperty(property);
        return prop?.GetValue(obj);
    }
}

// ─────────────────────────────────────────────
// Attribute for declarative state declaration
// ─────────────────────────────────────────────

// Host formats can use these attributes on C# classes
// to declare Rune bindings without XML/markup.

[AttributeUsage(AttributeTargets.Property)]
public class RuneStateAttribute : Attribute { }

[AttributeUsage(AttributeTargets.Property)]
public class RuneComputedAttribute : Attribute { }

[AttributeUsage(AttributeTargets.Method)]
public class RuneActionAttribute(string name) : Attribute
{
    public string Name { get; } = name;
}

[AttributeUsage(AttributeTargets.Property | AttributeTargets.Method)]
public class RuneIntentAttribute(string annotation) : Attribute
{
    public string Annotation { get; } = annotation;
}

// ─────────────────────────────────────────────
// Host Builder — reflection-based registration
// Scans a class for Rune attributes and auto-registers
// ─────────────────────────────────────────────

public static class RuneHostBuilder
{
    public static RuneHost From<T>(T instance) where T : class
    {
        var host = new RuneHost();
        var type = typeof(T);

        // Register state and computed properties
        foreach (var prop in type.GetProperties())
        {
            var statAttr    = prop.GetCustomAttribute<RuneStateAttribute>();
            var compAttr    = prop.GetCustomAttribute<RuneComputedAttribute>();
            var intentAttr  = prop.GetCustomAttribute<RuneIntentAttribute>();

            var name = ToKebabCase(prop.Name);

            if (compAttr is not null)
            {
                host.State.DeclareComputed(name, () => prop.GetValue(instance));
            }
            else if (statAttr is not null)
            {
                host.State.Declare(name, prop.GetValue(instance));

                // Wire ~ sync back: when host.Sync(name, value) is called,
                // write back to the property
                // (In a real implementation this uses INotifyPropertyChanged
                //  or a generated proxy — simplified here for clarity)
            }

            if (intentAttr is not null)
                host.RecordIntent(name, intentAttr.Annotation);
        }

        // Register action methods
        foreach (var method in type.GetMethods())
        {
            var actionAttr = method.GetCustomAttribute<RuneActionAttribute>();
            if (actionAttr is null) continue;

            var parameters = method.GetParameters();

            if (parameters.Length == 0)
                host.Actions.Register(actionAttr.Name,
                    () => { method.Invoke(instance, null); });
            else
                host.Actions.Register(actionAttr.Name,
                    (object?[] args) => { method.Invoke(instance, args); return Task.CompletedTask; });
        }

        return host;
    }

    private static string ToKebabCase(string name)
    {
        // "NewTask" → "new-task", "UnreadCount" → "unread-count"
        return string.Concat(name.Select((c, i) =>
            i > 0 && char.IsUpper(c) ? $"-{char.ToLower(c)}" : char.ToLower(c).ToString()));
    }
}
